pypesto/result/sample.py

```diff
@@ -5,19 +5,90 @@
 import numpy as np
 
 
-class McmcPtResult(dict):
+class SampleResultBase(dict):
+    """Result of the sample() function."""
+
+    def __init__(
+        self,
+        trace_x: np.ndarray = None,
+        trace_neglogpost: np.ndarray = None,
+        trace_neglogprior: np.ndarray = None,
+        burn_in: int = None,
+        time: float = 0.0,
+        auto_correlation: float = None,
+        effective_sample_size: float = None,
+        message: str = None,
+        **kwargs,
+    ):
+
+        super().__init__()
+
+        self.trace_x = trace_x
+        self.trace_neglogpost = trace_neglogpost
+        self.trace_neglogprior = trace_neglogprior
+        self.burn_in = burn_in
+        self.time = time
+        self.auto_correlation = auto_correlation
+        self.effective_sample_size = effective_sample_size
+        self.message = message
+
+    @staticmethod
+    def _check_trace_dimensions(trace_x, trace_neglogpost, trace_neglogprior):
+        raise NotImplementedError(
+            'Base class does not implement trace checks.'
+        )
+
+    def _get_n_samples(self):
+        """Return number of samples."""
+        return 0 if (self.trace_x is None) else self.trace_x.shape[0]
+
+    def __getattr__(self, key):
+        """Allow usage of keys like attributes."""
+        if key == 'n_samples':
+            return self._get_n_samples()
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(key)
+
+    def __setattr__(self, key, value):
+        """Allow usage of keys like attributes."""
+        if key == 'n_samples':
+            raise ValueError(
+                "n_samples can not be set, as they are computed "
+                "from self.trace_x."
+            )
+        self[key] = value
+
+    def __iter__(self):
+        self._index_iter = self.burn_in
+        return self
+
+    def __next__(self):
+        # return next sample
+        if self._index_iter < self.n_samples:
+            sample = self.trace_x[self._index_iter, :]
```

Comment:
I am not sure whether this simple iteration makes sense -- what if one wants to iterate over function values? which will happen often.